x/evm/keeper/msg_server.go

```diff
@@ -0,0 +1,187 @@
+	emsg, err := server.getEVMMessage(ctx, tx)
+	if err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	res, err := server.applyEVMMessage(ctx, emsg, stateDB, gp)
+	if err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	if err := server.writeReceipt(ctx, tx, emsg, res); err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	if err := stateDB.Finalize(); err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	errStr := ""
+	if res.Err != nil {
+		errStr = res.Err.Error()
+	}
+
+	return &types.MsgEVMTransactionResponse{
+		GasUsed:    res.UsedGas,
+		ReturnData: res.ReturnData,
+		VmError:    errStr,
+		Hash:       tx.Hash().Hex(),
+	}, nil
+}
+
+func (server msgServer) getGasPool(ctx sdk.Context) (sdk.Context, core.GasPool) {
+	if ctx.BlockGasMeter() == nil {
+		ctx = ctx.WithBlockGasMeter(sdk.NewInfiniteGasMeter())
+	}
+	if ctx.BlockGasMeter().Limit() == 0 {
+		// infinite gas meter
+		return ctx, math.MaxUint64
+	} else {
+		return ctx, core.GasPool(ctx.BlockGasMeter().Limit() - ctx.BlockGasMeter().GasConsumedToLimit())
+	}
+}
+
+func (server msgServer) getEVMMessage(ctx sdk.Context, tx *ethtypes.Transaction) (*core.Message, error) {
+	cfg := server.GetChainConfig(ctx).EthereumConfig(server.ChainID())
+	signer := ethtypes.MakeSigner(cfg, big.NewInt(ctx.BlockHeight()), uint64(ctx.BlockTime().Unix()))
+	return core.TransactionToMessage(tx, signer, server.getBaseFee())
+}
+
+func (server msgServer) applyEVMMessage(ctx sdk.Context, msg *core.Message, stateDB vm.StateDB, gp core.GasPool) (*core.ExecutionResult, error) {
+	coinbase, err := server.GetFeeCollectorAddress(ctx)
+	if err != nil {
+		return nil, err
+	}
+	cfg := server.GetChainConfig(ctx).EthereumConfig(server.ChainID())
+	blockCtx := vm.BlockContext{
+		CanTransfer: core.CanTransfer,
+		Transfer:    core.Transfer,
+		GetHash:     server.getHashFn(ctx),
+		Coinbase:    coinbase,
+		GasLimit:    gp.Gas(),
+		BlockNumber: big.NewInt(ctx.BlockHeight()),
+		Time:        uint64(ctx.BlockHeader().Time.Unix()),
+		Difficulty:  big.NewInt(0),       // only needed for PoW
+		BaseFee:     server.getBaseFee(), // feemarket not enabled
+		Random:      nil,                 // not supported
+	}
+	txCtx := core.NewEVMTxContext(msg)
+	evmInstance := vm.NewEVM(blockCtx, txCtx, stateDB, cfg, vm.Config{})
+	st := core.NewStateTransition(evmInstance, msg, &gp)
+	return st.TransitionDb()
+}
+
+func (server msgServer) writeReceipt(ctx sdk.Context, tx *ethtypes.Transaction, msg *core.Message, res *core.ExecutionResult) error {
+	var contractAddr common.Address
+	if msg.To == nil {
+		contractAddr = crypto.CreateAddress(msg.From, msg.Nonce)
+	} else if len(msg.Data) > 0 {
+		contractAddr = *msg.To
+	}
+
+	cumulativeGasUsed := res.UsedGas
+	if ctx.BlockGasMeter() != nil {
+		limit := ctx.BlockGasMeter().Limit()
+		cumulativeGasUsed += ctx.BlockGasMeter().GasConsumed()
+		if cumulativeGasUsed > limit {
+			cumulativeGasUsed = limit
```

Comment:
is this expected behavior? do we want to error out or something on situations where we are past the limit?

Reply:
it will eventually cause a panic here https://github.com/sei-protocol/sei-cosmos/blob/main/baseapp/baseapp.go#L1006C12-L1006C12. Here we want to still write cumulative gas to receipt (even if it will be reverted when the panic happens) because there might be other messages in the same tx (e.g. if the evm message is sent by a CW contract) that depend on this receipt